src/Microsoft.AspNetCore.Blazor.Razor.Extensions/BlazorDesignTimeNodeWriter.cs

```diff
@@ -0,0 +1,499 @@
+                    }
+                }
+                context.CodeWriter.WriteLine(";");
+            }
+        }
+
+        public override void WriteCSharpCodeAttributeValue(CodeRenderingContext context, CSharpCodeAttributeValueIntermediateNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            for (var i = 0; i < node.Children.Count; i++)
+            {
+                if (node.Children[i] is IntermediateToken token && token.IsCSharp)
+                {
+                    IDisposable linePragmaScope = null;
+                    var isWhitespaceStatement = string.IsNullOrWhiteSpace(token.Content);
+
+                    if (token.Source != null)
+                    {
+                        if (!isWhitespaceStatement)
+                        {
+                            linePragmaScope = context.CodeWriter.BuildLinePragma(token.Source.Value);
+                        }
+
+                        context.CodeWriter.WritePadding(0, token.Source.Value, context);
+                    }
+                    else if (isWhitespaceStatement)
+                    {
+                        // Don't write whitespace if there is no line mapping for it.
+                        continue;
+                    }
+
+                    context.AddSourceMappingFor(token);
+                    context.CodeWriter.Write(token.Content);
+
+                    if (linePragmaScope != null)
+                    {
+                        linePragmaScope.Dispose();
+                    }
+                    else
+                    {
+                        context.CodeWriter.WriteLine();
+                    }
+                }
+                else
+                {
+                    // There may be something else inside the statement like an extension node.
+                    context.RenderNode(node.Children[i]);
+                }
+            }
+        }
+
+        public override void WriteHtmlContent(CodeRenderingContext context, HtmlContentIntermediateNode node)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (node == null)
+            {
+                throw new ArgumentNullException(nameof(node));
+            }
+
+            // Do nothing
+        }
+
+        public override void BeginWriteAttribute(CodeWriter codeWriter, string key)
+        {
+            codeWriter
+                .WriteStartMethodInvocation($"{_scopeStack.BuilderVarName}.{nameof(BlazorApi.RenderTreeBuilder.AddAttribute)}")
+                .Write("-1")
+                .WriteParameterSeparator()
+                .WriteStringLiteral(key)
+                .WriteParameterSeparator();
```

Comment:
Out of interest, why do we need to emit the `AddAttribute` lines at design time?

Reply:
I left them out because overloading can make things somewhat confusing while you're typing.

In general an expression like `@(foo)` in Razor will be assigned to a variable of type `object` at design time, which really simplifies the space we need to reason about compared to `AddAttribute(...)` which is overloaded.  If we come up with cases where this illusion breaks we can try to figure something out.